File: src/main/java/club/xiaojiawei/utils/MouseUtil.java
package club.xiaojiawei.utils;

import club.xiaojiawei.dll.SystemDll;
import club.xiaojiawei.enums.ConfigurationEnum;
import com.sun.jna.platform.win32.WinDef;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.awt.*;
import java.util.Properties;
import java.util.concurrent.atomic.AtomicReference;

/**
 * 鼠标工具类
 * @author 肖嘉威
 * @date 2022/11/24 11:18
 */
@Slf4j
@Component
public class MouseUtil {

    private static Properties scriptConfiguration;

    private static AtomicReference<BooleanProperty> isPause = new AtomicReference<>(new SimpleBooleanProperty());

    public MouseUtil(Properties scriptConfiguration, AtomicReference<BooleanProperty> isPause) {
        MouseUtil.scriptConfiguration = scriptConfiguration;
        MouseUtil.isPause = isPause;
    }

    /**
     * 鼠标每次移动后的间隔时间：ms
     */
    private static final int MIN_MOVE_INTERVAL = 1;
    /**
     * 鼠标每次移动的距离：px
     */
    private static final int MOVE_DISTANCE = 10;

    @Getter
    private static final Point lastPoint = new Point(-1, -1);

    private static boolean validPoint(Point point) {
        if (point == null) {
            return false;
        }
        return point.x != -1 && point.y != -1;
    }

    public static void leftButtonClick(WinDef.HWND hwnd) {
        leftButtonClick(lastPoint, hwnd);
    }
    public static void leftButtonClick(Point pos, WinDef.HWND hwnd) {
        if (!isPause.get().get() && validPoint(pos)) {
            SystemDll.INSTANCE.leftClick(pos.x, pos.y, hwnd);
            savePos(pos);
        }
    }

    public static void rightButtonClick(WinDef.HWND hwnd) {
        rightButtonClick(lastPoint, hwnd);
    }
    public static void rightButtonClick(Point pos, WinDef.HWND hwnd) {
        if (!isPause.get().get() && validPoint(pos)) {
            SystemDll.INSTANCE.rightClick(pos.x, pos.y, hwnd);
            savePos(pos);
        }
    }


    public static void moveMouseByLine(Point endPos, WinDef.HWND hwnd) {
        moveMouseByLine(null, endPos, hwnd);
    }

    /**
     * 鼠标按照直线方式移动
     */
    public static void moveMouseByLine(Point startPos, Point endPos, WinDef.HWND hwnd) {
        if (!isPause.get().get() && validPoint(endPos)) {
            int endX = endPos.x;
            int endY = endPos.y;
            if (validPoint(startPos)) {
                int startX = startPos.x;
                int startY = startPos.y;
                if (Math.abs(startY - endY) <= 5) {
                    for (startX -= MOVE_DISTANCE; startX >= endX && !isPause.get().get(); startX -= MOVE_DISTANCE) {
                        SystemDll.INSTANCE.moveMouse(startX, startY, hwnd);
                        SystemUtil.delay(MIN_MOVE_INTERVAL, getMaxMoveInterval());
                    }
                } else if (Math.abs(startX - endX) <= 5) {
                    for (startY -= MOVE_DISTANCE; startY >= endY && !isPause.get().get(); startY -= MOVE_DISTANCE) {
                        SystemDll.INSTANCE.moveMouse(startX, startY, hwnd);
                        SystemUtil.delay(MIN_MOVE_INTERVAL, getMaxMoveInterval());
                    }
                } else {
                    double k = calcK(startX, startY, endX, endY);
                    double b = startY - k * startX;
                    for (startY -= MOVE_DISTANCE; startY >= endY && !isPause.get().get(); startY -= MOVE_DISTANCE) {
                        SystemDll.INSTANCE.moveMouse((int) ((startY - b) / k), startY, hwnd);
                        SystemUtil.delay(MIN_MOVE_INTERVAL, getMaxMoveInterval());
                    }
                }
            }
            SystemDll.INSTANCE.moveMouse(endX, endY, hwnd);
            savePos(endPos);
        }
    }

    private static void savePos(Point pos) {
        lastPoint.x = pos.x;
        lastPoint.y = pos.y;
    }

    /**
     * 计算斜率
     * @return double 斜率
     */
    private static double calcK(int startX, int startY, int endX, int endY) {
        return (double) (startY - endY) / (startX - endX);
    }


    private static int getMaxMoveInterval() {
        if (scriptConfiguration == null) {
            return Integer.parseInt(ConfigurationEnum.MOUSE_MOVE_INTERVAL.getDefaultValue());
        }
        return Integer.parseInt(scriptConfiguration.getProperty(ConfigurationEnum.MOUSE_MOVE_INTERVAL.getKey(), ConfigurationEnum.MOUSE_MOVE_INTERVAL.getDefaultValue()));
    }

}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/utils/MouseUtil.kt
package club.xiaojiawei.hsscript.utils

import club.xiaojiawei.enums.ModeEnum
import club.xiaojiawei.hsscript.bean.isDiscoverCardThread
import club.xiaojiawei.hsscript.bean.isOutCardThread
import club.xiaojiawei.hsscript.bean.single.WarEx
import club.xiaojiawei.hsscript.dll.CSystemDll
import club.xiaojiawei.hsscript.enums.ConfigEnum
import club.xiaojiawei.hsscript.initializer.DRIVER_LOCK
import club.xiaojiawei.hsscript.listener.WorkListener
import club.xiaojiawei.hsscript.status.Mode
import com.sun.jna.platform.win32.WinDef.HWND
import java.awt.Point

/**
 * 鼠标工具类
 * @author 肖嘉威
 * @date 2022/11/24 11:18
 */
object MouseUtil {

    private val mouseMovePauseStep: Int
        get() {
            return ConfigUtil.getInt(ConfigEnum.PAUSE_STEP)
        }

    private val prevPoint = Point(0, 0)

    private fun validatePoint(point: Point?): Boolean {
        return point?.let {
            it.x != -1 && it.y != -1
        } == true
    }

    private fun savePos(pos: Point) {
        prevPoint.x = pos.x
        prevPoint.y = pos.y
    }

    /**
     * 计算斜率
     * @return double 斜率
     */
    private fun calcK(startX: Int, startY: Int, endX: Int, endY: Int): Double {
        return (startY - endY).toDouble() / (startX - endX)
    }

    fun leftButtonClick(hwnd: HWND?) {
        leftButtonClick(prevPoint, hwnd)
    }

    fun leftButtonClick(
        pos: Point,
        hwnd: HWND?,
        mouseMode: Int = ConfigExUtil.getMouseControlMode().code
    ) {
        if (!validateEnv(hwnd)) return
        if (validatePoint(pos)) {
            DRIVER_LOCK.lock()
            try {
                if (!WorkListener.working) return

                if (prevPoint != pos) {
                    CSystemDll.INSTANCE.simulateHumanMoveMouse(
                        prevPoint.x,
                        prevPoint.y,
                        pos.x,
                        pos.y,
                        hwnd,
                        mouseMovePauseStep,
                        mouseMode
                    )
                    SystemUtil.delayShort()
                }
                CSystemDll.INSTANCE.leftClick(pos.x.toLong(), pos.y.toLong(), hwnd, mouseMode)
                savePos(pos)
            } finally {
                DRIVER_LOCK.unlock()
            }
        }
    }

    fun rightButtonClick(hwnd: HWND?) {
        rightButtonClick(prevPoint, hwnd)
    }

    fun rightButtonClick(
        pos: Point,
        hwnd: HWND?,
        mouseMode: Int = ConfigExUtil.getMouseControlMode().code
    ) {
        if (!validateEnv(hwnd)) return

        if (validatePoint(pos)) {
            DRIVER_LOCK.lock()
            try {
                if (!WorkListener.working && Mode.currMode !== ModeEnum.GAMEPLAY) return

                if (prevPoint != pos) {
                    CSystemDll.INSTANCE.simulateHumanMoveMouse(
                        prevPoint.x,
                        prevPoint.y,
                        pos.x,
                        pos.y,
                        hwnd,
                        mouseMovePauseStep,
                        mouseMode
                    )
                    SystemUtil.delayShort()
                }
                CSystemDll.INSTANCE.rightClick(pos.x.toLong(), pos.y.toLong(), hwnd, mouseMode)
                savePos(pos)
            } finally {
                DRIVER_LOCK.unlock()
            }
        }
    }

    fun moveMouseByHuman(endPos: Point, hwnd: HWND?) {
        moveMouseByHuman(null, endPos, hwnd)
    }

    private fun validateEnv(hwnd: HWND?): Boolean {
//        选择卡牌时间只让特定线程执行
        if (WarEx.war.isChooseCardTime && !isDiscoverCardThread()) return false
        hwnd ?: return false
        return !(ConfigUtil.getBoolean(ConfigEnum.DISABLE_MOUSE) ||
                ((isOutCardThread() || isDiscoverCardThread()) && Mode.currMode !== ModeEnum.GAMEPLAY)
                || !WorkListener.working)
    }

    /**
     * 鼠标移动
     */
    fun moveMouseByHuman(
        startPos: Point?,
        endPos: Point,
        hwnd: HWND?,
        mouseMode: Int = ConfigExUtil.getMouseControlMode().code
    ) {
        if (!validateEnv(hwnd)) return

        DRIVER_LOCK.lock()
        try {
            if (!WorkListener.working) return

            val prevPoint = prevPoint
            if (validatePoint(startPos)) {
                startPos!!
                moveMouseByHuman(startPos, hwnd)
                if (validatePoint(endPos)) {
                    SystemUtil.delayShort()
                    if (startPos != endPos) {
                        CSystemDll.INSTANCE.simulateHumanMoveMouse(
                            startPos.x,
                            startPos.y,
                            endPos.x,
                            endPos.y,
                            hwnd,
                            mouseMovePauseStep,
                            mouseMode
                        )
                        savePos(endPos)
                    }
                }
            } else if (validatePoint(prevPoint)) {
                if (prevPoint != endPos) {
                    CSystemDll.INSTANCE.simulateHumanMoveMouse(
                        prevPoint.x,
                        prevPoint.y,
                        endPos.x,
                        endPos.y,
                        hwnd,
                        mouseMovePauseStep,
                        mouseMode
                    )
                    savePos(endPos)
                }
            }
        } finally {
            DRIVER_LOCK.unlock()
        }
    }

}
